File: kf_pom/kf_pom/custom_script/material_request/material_request.js
frappe.ui.form.on("Material Request",{
	refresh:function(frm){
        $.each(frm.doc.items, function(i,v) {
              frappe.model.set_value(v.doctype, v.name, "schedule_date", frm.doc.schedule_date);                           
              frappe.item_quantity += v.qty
        });
		if(frm.is_new()) {
            frm.set_value('requestor_email',frappe.session.user_email);
            frm.set_value('requestor_name',frappe.session.user_fullname);
        } 
		
		cur_frm.fields_dict['sub_category'].get_query = function(doc, cdt, cdn) {
		    return {
		        filters: [
		            		['Sub Category','category', '=', frm.doc.category]
						]
			}
		}
        if(frappe.user.has_role('Requestor/Site Manager')) {
            frm.set_df_property('approver_comments','read_only',1)
        }
        if((frm.doc.workflow_state=="Rejected by Procurement Approver"||frm.doc.workflow_state=="Submitted by Requestor") && frappe.user.has_role('Commercial Approver')) {
            frm.set_df_property('approver_comments','reqd',1)
        }
        if(frm.is_new()){
            frm.set_value("tc_name",'Standard Template')
            frm.set_value("kf_contact_name","Sarvesh Tiwari")
            frm.set_value("kf_contact_email","sarvesh.tiwari@in.knightfrank.com")
            frm.set_value("kf_contact_no","8291900219")
        }
	},
    category: function(frm) {
        if(frm.doc.category){
            frm.set_value('sub_category','')
            cur_frm.fields_dict['sub_category'].get_query = function(doc, cdt, cdn) {
                return {
                    filters: [
                                ['Sub Category','category', '=', frm.doc.category]
                            ]
                }
            } 
        } else {
            frm.set_value('sub_category','')
        }
    },
    kf_contact_email: function(frm) {
        if(frm.doc.kf_contact_email) {
            frappe.call({
                method: 'frappe.client.get_value',
                args: {
                    'doctype': 'User',
                    'filters': {'name': frm.doc.kf_contact_email},
                    'fieldname': [
                        'mobile_no',
                        'full_name'
                    ]
                },
                callback: function(r) {
                    if (!r.exc) {
                        // code snippet
                        if(r.message){
                            frm.set_value("kf_contact_name",r.message.full_name)
                            frm.set_value("kf_contact_no",r.message.mobile_no)
                        }
                    }
                }
            });
        } else {
            frm.set_value("kf_contact_name","")
            frm.set_value("kf_contact_no","")
        }
    },
	onload:function(frm) {
        var item_quantity = 0
        $.each(frm.doc.items, function(i,v) {                       
              item_quantity += v.qty
        });
        frappe.call({
            method: 'kf_pom.kf_pom.custom_script.material_request.material_request.check_po',
            args: {
                'mi': frm.doc.name
            },
            callback: function(r) {
                if(r.message)
                {
                    if(r.message[0].tot_qty == item_quantity)
                    {
                        frm.remove_custom_button(__("Create"))
                    }
                }
            }
        });
        if(frm.is_new()) {
            frm.set_value('requestor_email',frappe.session.user_email);
            frm.set_value('requestor_name',frappe.session.user_fullname);
            frm.set_value('schedule_date',frappe.datetime.add_days(frappe.datetime.get_today(), 3))
        }
        if(frm.is_new()){
            frm.set_value("tc_name",'Standard Template')
            frm.set_value("kf_contact_name","Sarvesh Tiwari")
            frm.set_value("kf_contact_email","sarvesh.tiwari@in.knightfrank.com")
            frm.set_value("kf_contact_no","8291900219")
        }
        frm.set_query('kf_customer_shipping_address', function(doc) {
            if(!doc.kf_customer) {
                frappe.throw(_('Please select Customer'));
            }

            return {
                // query: 'frappe.contacts.doctype.address.address.address_query',
                query: 'kf_pom.kf_pom.custom_script.material_request.material_request.address_query',
                filters: {
                    link_doctype: 'Customer',
                    link_name: doc.kf_customer
                }
            };
        });
    },
    company: function(frm) {
        frm.set_query('billing_address', function(doc) {

            return {
                // query: 'frappe.contacts.doctype.address.address.address_query',
                query: 'kf_pom.kf_pom.custom_script.material_request.material_request.address_query',
                filters: {
                    link_doctype: 'Company',
                    link_name: doc.company
                }
            };
        });
    },
    billing_address:function(frm) {
        if(frm.doc.billing_address) {
            frm.set_value('company_billing_add',frm.doc.billing_address)
        } else {
            frm.set_value('company_billing_add','')
        }
    },
    kf_customer: function(frm){
        //to filter customer addresses in drop-down
        frm.set_value("kf_customer_shipping_address","")
        
        frm.set_query('kf_customer_shipping_address', function(doc) {
            if(!doc.kf_customer) {
                frappe.throw(_('Please select Customer'));
            }

            return {
                // query: 'frappe.contacts.doctype.address.address.address_query',
                query: 'kf_pom.kf_pom.custom_script.material_request.material_request.address_query',
                filters: {
                    link_doctype: 'Customer',
                    link_name: doc.kf_customer
                }
            };
        });
    },
    kf_customer_shipping_address: function(frm) {
        //to fetch customer address on PO
        if(frm.doc.kf_customer_shipping_address) {
            frappe.call({
                method: "frappe.contacts.doctype.address.address.get_address_display",
                args: {"address_dict": frm.doc.kf_customer_shipping_address },
                callback: function(r) {
                    if(r.message) {
                        console.log(r.message)
                        frm.set_value("kf_customer_shipping_address_display", r.message)
                    }
                }
            })
        }

        if(frm.doc.kf_customer_shipping_address) {
            frappe.call({
                method: "kf_pom.kf_pom.custom_script.material_request.material_request.get_address_details",
                args: {"name": frm.doc.kf_customer_shipping_address },
                callback: function(r) {
                    if(r.message) {
                        frm.set_value('kf_sub_location',r.message[0].sub_location)
                        frm.set_value('commercial_approver',r.message[0].commercial_approver)
                    }
                }
            })
        } else {
            frm.set_value("kf_customer_shipping_address_display", "");
            frm.set_value('kf_sub_location',"")
            frm.set_value('commercial_approver',"")
        }
        
    },
    requestor_email: function(frm) {
        //set requestor mobile no
        if(frm.doc.requestor_email){
            frappe.db.get_value('User',frm.doc.requestor_email,'mobile_no')
            .then(r => {
                if(r.message){
                    frm.set_value('requestor_contact_no',r.message.mobile_no)
                }
            })
        }
    },
    validate:function(frm){
        var from_date=new Date(frm.doc.po_validity_from_date);
        var to_date=new Date(frm.doc.po_validity_to_date);
    	if (from_date && to_date){
    		if (from_date>to_date){
    			frappe.throw("From Date Should not Exceed To Date")
    		}
    	}
    }
});

File: kf_pom/kf_pom/custom_script/material_request/material_request.py
from __future__ import unicode_literals
from frappe.model.document import Document
import frappe
import json
from frappe.utils.user import get_user_fullname

def validate(doc,method=None):
	# update amount for all the entries in items table and calculate the grand_total
	if(len(doc.items)>0):
		grand_total = 0.00
		for item in doc.items:
			amount = float(item.rate)*item.qty
			item.amount = amount
			grand_total= grand_total + amount
		doc.grand_total = grand_total

	if(doc.commercial_approver == None or doc.commercial_approver == ''):
		frappe.throw("Please get the commercial_approver mapped to the customer before creating the Purchase requisition for this customer")

	# validations for comments to be added by approvers
	if((doc.workflow_state == 'Commercial Approver Approved' or doc.workflow_state ==  'Rejected by Commercial Approver' )and 'Commercial Approver' in frappe.get_roles()) :
		if doc.	approver_comments == None or doc.approver_comments == '':
			frappe.throw("Approver's comments are mandatory")
		else:
			doc.add_comment('Comment', text=doc.approver_comments)
			doc.approver_comments = ''

	if(doc.workflow_state == 'Rejected by Procurement Approver' and 'Procurement  Approver' in frappe.get_roles()):
		if doc.	approver_comments == None or doc.approver_comments == '':
			frappe.throw("Approver's comments are mandatory")
		else:
			doc.add_comment('Comment', text=doc.approver_comments)
			doc.approver_comments = ''

	if(doc.workflow_state == 'Procurement Approver Approved' and 'Procurement  Approver' in frappe.get_roles()):
		if doc.approver_comments:
			doc.add_comment('Comment', text=doc.approver_comments)
			doc.approver_comments = ''

	from frappe.utils import get_url	
	base_url_dev = get_url() + "/app" 

	# send the emails on creation,approval and rejection
	if 'Requestor/Site Manager' in frappe.get_roles() and doc.workflow_state == 'Submitted by Requestor':
		s_email = frappe.db.get_singles_dict("KF Email Settings")
		r_email = doc.commercial_approver
		cc_email = doc.requestor_email
		url = base_url_dev + "/material-request/" + doc.name
		subject = """Purchase Requisition %s for %s has been submitted for your Approval by %s """%(doc.name,doc.kf_customer,doc.requestor_name)
		msg = """Hello %s, <br> The Purchase Requisition %s (Category: %s, Sub category: %s) for %s has been submitted for your approval by %s
		<br> Link : %s <br><br>Thanks,<br>Knight Frank Procurement Team
		"""%(get_user_fullname(r_email),doc.name,doc.category,doc.sub_category,doc.kf_customer,doc.requestor_name,url)
		if s_email["email_configuration"] == "1":
			frappe.sendmail(recipients=r_email,cc=cc_email,subject= subject,content=msg)

	if 'Commercial Approver' in frappe.get_roles() and doc.workflow_state == 'Commercial Approver Approved':
		cc_email = doc.requestor_email 
		s_email = frappe.db.get_singles_dict("KF Email Settings")
		r_email = s_email["procurement_approver"]
		url = base_url_dev + "/material-request/" + doc.name
		subject = """Purchase Requisition %s for %s has been submitted for your Approval by %s"""%(doc.name,doc.kf_customer,get_user_fullname(doc.modified_by))
		msg = """Hello %s, <br> The Purchase Requisition %s (Category: %s, Sub category: %s) for %s has been submitted for your approval by %s.
		<br> Link: %s <br><br>Thanks,<br>Knight Frank Procurement Team
		"""%(get_user_fullname(r_email),doc.name,doc.category,doc.sub_category,doc.kf_customer,get_user_fullname(doc.modified_by),url)
		if s_email["email_configuration"] == "1":
			frappe.sendmail(recipients=r_email,cc=cc_email,subject=subject,content=msg)

	if 'Commercial Approver' in frappe.get_roles() and doc.workflow_state == 'Rejected by Commercial Approver':
		s_email = frappe.db.get_singles_dict("KF Email Settings")
		r_email = doc.requestor_email
		url = base_url_dev + "/material-request/" + doc.name
		comments = frappe.db.sql(''' 
			Select content from `tabComment` where reference_doctype='Material Request' 
			and reference_name=%s 
			and comment_type='comment'
			and owner = %s
			''',(doc.name,doc.modified_by),as_list=1)
		subject = """Purchase Requisition %s for %s has been Rejected by %s"""%(doc.name,doc.kf_customer,get_user_fullname(doc.modified_by))
		msg = """Hello %s, <br> The Purchase Requisition %s (Category: %s, Sub category: %s) for %s has been rejected by %s with the following Comments: %s
		<br>Link: %s <br><br>Thanks,<br>Knight Frank Procurement Team
		"""%(get_user_fullname(r_email),doc.name,doc.category,doc.sub_category,doc.kf_customer,get_user_fullname(doc.modified_by),comments[0][0],url)
		if s_email["email_configuration"] == "1":
			frappe.sendmail(recipients=r_email,subject=subject,content=msg)	

	if 'Procurement  Approver' in frappe.get_roles() and doc.workflow_state == 'Procurement Approver Approved':
		s_email = frappe.db.get_singles_dict("KF Email Settings")
		r_email = doc.requestor_email
		url = base_url_dev + "/material-request/" + doc.name
		subject="""Purchase Requisition %s for %s has been submitted for your Approval by %s"""%(doc.name,doc.kf_customer,get_user_fullname(doc.modified_by))
		msg = """Hello %s, <br> The Purchase Requisition %s (Category: %s, Sub category: %s) for %s has been Approved by %s
		<br>Link: %s <br><br>Thanks,<br>Knight Frank Procurement Team
		"""%(get_user_fullname(r_email),doc.name,doc.category,doc.sub_category,doc.kf_customer,get_user_fullname(doc.modified_by),url)
		if s_email["email_configuration"] == "1":
			frappe.sendmail(recipients=r_email,subject=subject,content=msg)	

	if 'Procurement  Approver' in frappe.get_roles() and doc.workflow_state == 'Rejected by Procurement Approver':
		s_email = frappe.db.get_singles_dict("KF Email Settings")
		r_email = doc.requestor_email
		url = base_url_dev + "/material-request/" + doc.name
		comments = frappe.db.sql(''' 
			Select content from `tabComment` where reference_doctype='Material Request' 
			and reference_name=%s 
			and comment_type='comment'
			and owner = %s
			''',(doc.name,doc.modified_by),as_list=1)
		subject = """Purchase Requisition %s for %s has been Rejected by %s"""%(doc.name,doc.kf_customer,get_user_fullname(doc.modified_by))
		msg = """Hello %s, <br> The Purchase Requisition %s (Category: %s, Sub category: %s) for %s has been rejected by %s with the following Comments : %s.
		<br>Link: %s <br><br>Thanks,<br>Knight Frank Procurement Team
		"""%(get_user_fullname(r_email),doc.name,doc.category,doc.sub_category,doc.kf_customer,get_user_fullname(doc.modified_by),comments[0][0],url)
		if s_email["email_configuration"] == "1":
			frappe.sendmail(recipients=r_email,subject=subject,content=msg)

def get_permission_query_conditions(doctype):
	#Commercial Approver can see only the MR for himself
	if frappe.session.user == "Administrator":
		return ""

	if "Director Approver" in frappe.get_roles() or "Procurement  Approver" in frappe.get_roles() or "Vendor" in frappe.get_roles():
		names = frappe.db.sql("""select name from `tabMaterial Request` 
								where workflow_state not in ('Created by Requestor') """)

		if names:
			names = ",".join("'" + i[0]+"'" for i in names)
			names = "(" + names + ")"

		if names:
			return """(`tabMaterial Request`.name in %s)"""%(names)

		if not names:
		#to return nothing
			return """(`tabMaterial Request`.name is NULL)"""

	if "Commercial Approver" in frappe.get_roles():
		user = frappe.session.user
		names = frappe.db.sql("""select name from `tabMaterial Request` 
								where commercial_approver=%s 
								and workflow_state not in ('Created by Requestor') """,(user))

		if names:
			names = ",".join("'" + i[0]+"'" for i in names)
			names = "(" + names + ")"

		if names:
			return """(`tabMaterial Request`.name in %s)"""%(names)

		if not names:
		#to return nothing
			return """(`tabMaterial Request`.name is NULL)"""


 # To include the addres_line1 in address drop-down on PR
@frappe.whitelist()
@frappe.validate_and_sanitize_search_inputs
def address_query(doctype, txt, searchfield, start, page_len, filters):
	from frappe.desk.reportview import get_match_cond

	link_doctype = filters.pop('link_doctype')
	link_name = filters.pop('link_name')

	return frappe.db.sql("""select `tabAddress`.name, `tabAddress`.address_line1,`tabAddress`.city
					from 
						`tabAddress`,`tabDynamic Link` 
					where
						`tabDynamic Link`.parent = `tabAddress`.name and
						`tabDynamic Link`.parenttype = 'Address' and
						`tabDynamic Link`.link_doctype = %(link_doctype)s and 
						`tabDynamic Link`.link_name = %(link_name)s and
						ifnull(`tabAddress`.disabled, 0) = 0
						{mcond}
						order by
						if(locate(%(_txt)s, `tabAddress`.address_line1), locate(%(_txt)s, `tabAddress`.address_line1), 99999),
						`tabAddress`.idx desc, `tabAddress`.address_line1
						limit %(start)s, %(page_len)s""".format(
							mcond=get_match_cond(doctype),
							key=searchfield
							), {
							'txt': '%' + txt + '%',
							'_txt': txt.replace("%", ""),
							'start': start,
							'page_len': page_len,
							'link_name': link_name,
							'link_doctype': link_doctype
						})